tests/Security/TwoFactor/Validator/Constraints/UserGoogleTotpCodeTest.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Tests\Security\TwoFactor\Validator\Constraints;
+
+use Generator;
+use PHPUnit\Framework\Attributes\DataProvider;
+use PHPUnit\Framework\Attributes\Test;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Validator\Constraints\UserGoogleTotpCode;
+use Scheb\TwoFactorBundle\Tests\TestCase;
+use Symfony\Component\Validator\Mapping\ClassMetadata;
+use Symfony\Component\Validator\Mapping\Loader\AttributeLoader;
+
+class UserGoogleTotpCodeTest extends TestCase
+{
+    #[Test]
+    public function validatedBy_objectInitialized_returnDefault(): void
+    {
+        $constraint = new UserGoogleTotpCode();
+
+        $this->assertSame('scheb_two_factor.security.totp.validator.user_google_totp_code', $constraint->validatedBy());
+    }
+
+    #[Test]
+    #[DataProvider('provideServiceValidatedConstraints')]
+    public function validatedBy_customValue_returnValid(UserGoogleTotpCode $constraint): void
+    {
+        $this->assertSame('my_service', $constraint->validatedBy());
+    }
+
+    /**
+     * @return Generator<string, list{UserGoogleTotpCode}>
+     */
+    public static function provideServiceValidatedConstraints(): iterable
+    {
+        yield 'Doctrine style' => [new UserGoogleTotpCode(['service' => 'my_service'])];
+
+        yield 'named arguments' => [new UserGoogleTotpCode(service: 'my_service')];
+
+        $metadata = new ClassMetadata(UserTotpCodeDummy::class);
+        self::assertTrue((new AttributeLoader())->loadClassMetadata($metadata));
+
+        yield 'attribute' => [$metadata->getPropertyMetadata('b')[0]->getConstraints()[0]];
```

Comment:
This doesn't feel right to me, having an assert in a _data provider_, which should return _static_ values to represent different test scenarios. As soon as there's logic and asserts involved, the data provider is no longer a data provider, but becomes a test case.

I'd rather move the check for the "validatedBy" value to the `attributes_loadMetadata_returnValid` test. To my understanding, this is a "check values loaded from attributes are properly processed" test case, so would be better to assert the attribute-provided validatedBy value there.

(same in the UserTotpCodeTest)